File: java/src/leetcode/editor/cn/70-ClimbingStairs.java
package leetcode.editor.cn;

/*
 * @lc app=leetcode.cn id=70 lang=java
 *
 * [70] Climbing Stairs
 *
 * https://leetcode-cn.com/problems/climbing-stairs/description/
 *
 * algorithms
 * Easy (53.46%)
 * Likes:    2211
 * Dislikes: 0
 * Total Accepted:    710.1K
 * Total Submissions: 1.3M
 * Testcase Example:  '2'
 *
 * You are climbing a staircase. It takes n steps to reach the top.
 * 
 * Each time you can either climb 1 or 2 steps. In how many distinct ways can
 * you climb to the top?
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: n = 2
 * Output: 2
 * Explanation: There are two ways to climb to the top.
 * 1. 1 step + 1 step
 * 2. 2 steps
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: n = 3
 * Output: 3
 * Explanation: There are three ways to climb to the top.
 * 1. 1 step + 1 step + 1 step
 * 2. 1 step + 2 steps
 * 3. 2 steps + 1 step
 * 
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= n <= 45
 * 
 * 
 */

// @lc code=start
class Solution {
    /// 备忘录
    private int[] mem;

    public int climbStairs(int n) {
        mem = new int[n + 1];
        return f(n);
    }

    private int f(int n) {
        if (n == 1) {
            return 1;
        }
        if (n == 2) {
            return 2;
        }
        if (mem[n] != 0)
            return mem[n];
        mem[n] = f(n - 1) + f(n - 2);
        return mem[n];
    }
}
// @lc code=end

File: javascript/src/leetcode/editor/cn/70-ClimbingStairs.js
/*
 * @lc app=leetcode.cn id=70 lang=javascript
 *
 * [70] 爬楼梯
 *
 * https://leetcode-cn.com/problems/climbing-stairs/description/
 *
 * algorithms
 * Easy (49.81%)
 * Likes:    1100
 * Dislikes: 0
 * Total Accepted:    231.3K
 * Total Submissions: 463.9K
 * Testcase Example:  '2'
 *
 * 假设你正在爬楼梯。需要 n 阶你才能到达楼顶。
 * 
 * 每次你可以爬 1 或 2 个台阶。你有多少种不同的方法可以爬到楼顶呢？
 * 
 * 注意：给定 n 是一个正整数。
 * 
 * 示例 1：
 * 
 * 输入： 2
 * 输出： 2
 * 解释： 有两种方法可以爬到楼顶。
 * 1.  1 阶 + 1 阶
 * 2.  2 阶
 * 
 * 示例 2：
 * 
 * 输入： 3
 * 输出： 3
 * 解释： 有三种方法可以爬到楼顶。
 * 1.  1 阶 + 1 阶 + 1 阶
 * 2.  1 阶 + 2 阶
 * 3.  2 阶 + 1 阶
 * 
 * 
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function (n) {
    return solution2(n)
};
// @lc code=end

var solution0 = function (n) {
    var data = []
    data[0] = 1
    data[1] = 1
    for (var i = 2; i <= n; i++) {
        data[i] = data[i - 1] + data[i - 2]
    }

    return data[n]
}

/// 递归超时
var solution1 = function (n) {
    if (n == 0 || n == 1) {
        return 1;
    }

    return solution1(n - 1) + solution1(n - 2)
}

var solution2 = function (n) {
    var p = 0
    var q = 0
    var r = 1

    for (var i = 1; i <= n; i++) {
        p = q;
        q = r;
        r = p + q;
    }

    return r
}